### ISSUE TITLE
[Error: SFTP session ended early]

### ISSUE THREAD

**Issue by nickgzzjr** [mscdex/ssh2-streams#28]
I keep getting this error. I am actually using gulp-ssh, but the problem ended up happening in one of the files here.

```
[Error: SFTP session ended early]
_stream_writable.js:313
    cb(er);
    ^
TypeError: object is not a function
    at onwriteError (_stream_writable.js:313:5)
    at onwrite (_stream_writable.js:335:5)
    at WritableState.onwrite (_stream_writable.js:105:5)
    at /Users/nico/Work/nico/node_modules/gulp-ssh/node_modules/ssh2/node_modules/ssh2-streams/lib/sftp.js:2753:14
    at state.requests.(anonymous function).cb (/Users/nico/Work/nico/node_modules/gulp-ssh/node_modules/ssh2/node_modules/ssh2-streams/lib/sftp.js:961:15)
    at SFTPStream._cleanup (/Users/nico/Work/nico/node_modules/gulp-ssh/node_modules/ssh2/node_modules/ssh2-streams/lib/sftp.js:191:38)
    at SFTPStream.end (/Users/nico/Work/nico/node_modules/gulp-ssh/node_modules/ssh2/node_modules/ssh2-streams/lib/sftp.js:160:8)
    at SFTPWrapper.end (/Users/nico/Work/nico/node_modules/gulp-ssh/node_modules/ssh2/lib/SFTPWrapper.js:29:23)
    at end (/Users/nico/Work/nico/node_modules/gulp-ssh/index.js:253:18)
    at DestroyableTransform._flush (/Users/nico/Work/nico/node_modules/gulp-ssh/index.js:286:5)
```

After playing around with it for a few hours I found that if I comment out 

```
SFTPStream.prototype._cleanup = function(callback) {
  var state = this._state;

  state.pktBuf = undefined; // give GC something to do

  var requests = state.requests,
      keys = Object.keys(requests),
      len = keys.length;
  if (len) {
    var err = new Error('SFTP session ended early');
    for (var i = 0, cb; i < len; ++i)
      (cb = requests[keys[i]].cb) && cb(err);
  }

  if (this.readable)
    this.push(null);
  if (callback !== false) {
    this.debug('DEBUG[SFTP]: Parser: Malformed packet');
    callback && callback(new Error('Malformed packet'));
  }
};
```

as such

```
SFTPStream.prototype._cleanup = function(callback) {
  var state = this._state;

  state.pktBuf = undefined; // give GC something to do

  var requests = state.requests,
      keys = Object.keys(requests),
      len = keys.length;
  if (len) {
    var err = new Error('SFTP session ended early');
    // for (var i = 0, cb; i < len; ++i)
      // (cb = requests[keys[i]].cb) && cb(err);
  }

  if (this.readable)
    this.push(null);
  if (callback !== false) {
    this.debug('DEBUG[SFTP]: Parser: Malformed packet');
    callback && callback(new Error('Malformed packet'));
  }
};
```

It works.

**Comment by DefunctExodus** [comment]
I have actually been having the exact same issue

**Comment by mscdex** [comment]
This should be fixed in the latest release.

**Comment by Rommero** [comment]
Me too. I also traced the problem till the cleanup function `SFTPStream.prototype._cleanup`. The `requests` variable seem to have a callback remaining assigned in `ReadStream.prototype.close`. 
Waiting for your support and wish you good luck in the following release

**Comment by mscdex** [comment]
@Rommero Can you try the master branch?

**Comment by Rommero** [comment]
@mscdex still fails =(

**Comment by zacronos** [comment]
I have encountered the same problem using ssh2@0.4.13 / ssh2-streams@0.0.21, with the stack trace also pointing to `SFTPStream.prototype._cleanup`.

In my case, I would connect the sshClient, start an SFTP session, use `sftp.createReadStream()`, pipe it into another stream, and when that other stream finishes I call `sshClient.end()`.

I will try creating a fork of ssh2 so I can set the ssh2-streams version to 0.0.14, and see if that fixes it as @Rommero suggests.

**Comment by zacronos** [comment]
Version 0.0.14 does seem to avoid this error for me, so for now I'm going to have to use my fork of ssh2 which locks the ssh2-streams version at 0.0.14.  Anyone else encountering this issue is free to use the fork as well -- I'll try to keep it up to date with the main ssh2 until this bug is fixed and we can safely return to the main fork.

https://github.com/realtymaps/ssh2

**Comment by mscdex** [comment]
This should be fixed in v0.0.22.
